java-core/google-cloud-core/src/test/java/com/google/cloud/TimestampTest.java

```diff
@@ -237,14 +238,79 @@ void parseTimestampWithoutTimeZoneOffset() {
+    assertThat(Timestamp.parseTimestampDuration("0001-01-01T00:00:00.000000001Z"))
+        .isEqualTo(Timestamp.ofTimeSecondsAndNanos(Timestamp.MIN_VALUE.getSeconds(), 1));
+    assertThat(Timestamp.parseTimestampDuration("9999-12-31T23:59:59.999999998Z"))
+        .isEqualTo(Timestamp.ofTimeSecondsAndNanos(Timestamp.MAX_VALUE.getSeconds(), 999999998));
+    // Common use cases
+    assertThat(Timestamp.parseTimestampDuration("2020-07-10T14:03:00.123-07:00"))
+        .isEqualTo(Timestamp.ofTimeSecondsAndNanos(1594414980, 123000000));
+    assertThat(Timestamp.parseTimestampDuration("2020-12-06T19:21:12.123+05:30"))
         .isEqualTo(Timestamp.ofTimeSecondsAndNanos(1607262672, 123000000));
-    assertThat(Timestamp.parseTimestamp("2020-07-10T14:03:00-07:00"))
+    // We also confirm that parsing a timestamp with nano precision will behave the same as the
+    // threeten counterpart
+    assertThat(Timestamp.parseTimestampDuration("2020-12-06T19:21:12.123+05:30"))
+        .isEqualTo(Timestamp.parseTimestamp("2020-12-06T19:21:12.123+05:30"));
+    // Timestamps with fractional seconds at nanosecond level
+    assertThat(Timestamp.parseTimestampDuration("2020-12-06T19:21:12.123456789+05:30"))
+        .isEqualTo(Timestamp.ofTimeSecondsAndNanos(1607262672, 123456789));
+    // Fractional seconds beyond nanos should throw an exception
+    assertThrows(
+        DateTimeParseException.class,
+        () -> Timestamp.parseTimestampDuration("2020-12-06T19:21:12.123456789321+05:30"));
+    // Missing components (should throw exceptions)
+    assertThrows(
+        DateTimeParseException.class, () -> Timestamp.parseTimestampDuration("2020-12-06"));
+    // Whitespace should not be supported
+    assertThrows(
+        DateTimeParseException.class,
+        () -> Timestamp.parseTimestampDuration("  2020-12-06T19:21:12.123+05:30  "));
+    // It should be case-insensitive
+    assertThat(Timestamp.parseTimestampDuration("2020-07-10t14:03:00-07:00"))
         .isEqualTo(Timestamp.ofTimeSecondsAndNanos(1594414980, 0));
+    // Invalid time zone offsets
+    assertThrows(
+        DateTimeParseException.class,
+        () -> Timestamp.parseTimestampDuration("2020-12-06T19:21:12.123+25:00"));
+    assertThrows(
+        DateTimeParseException.class,
+        () -> Timestamp.parseTimestampDuration("2020-12-06T19:21:12.123-25:00"));
+    // Int values for SecondOfMinute should be between 0 and 59
+    assertThrows(
+        DateTimeParseException.class,
+        () -> Timestamp.parseTimestampDuration("2016-12-31T23:59:60Z"));
+  }
+
+  @Test
+  void demoAssertion1() {
+    assertThat(Timestamp.parseTimestampDuration("2020-07-10T14:03:00.123-07:00"))
+        .isEqualTo(Timestamp.ofTimeSecondsAndNanos(1594414980, 123000000));
+  }
+
+  @Test
+  void demoAssertion2() {
+    assertThat(Timestamp.parseTimestampDuration("2020-12-06T19:21:12.123+05:30"))
+        .isEqualTo(Timestamp.ofTimeSecondsAndNanos(1607262672, 123000000));
+  }
```

Comment:
I added these two tests to separately confirm that both offset conversion test cases are failing when using `appendOffsetId()` using java.time in java 8

Reply:
```
Error:  Failures: 
Error:    TimestampTest.demoAssertion1:293 value of: parseTimestampDuration(...)
expected: 2020-07-10T21:03:00.123000000Z
but was : 2020-07-10T14:03:00.123000000Z
Error:    TimestampTest.demoAssertion2:299 value of: parseTimestampDuration(...)
expected: 2020-12-06T13:51:12.123000000Z
but was : 2020-12-06T19:21:12.123000000Z
```